components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthnMissingChallengeQuestionsHandler.java

```diff
@@ -0,0 +1,342 @@
+                    "for force challenge password ", e);
+            return null;
+        }
+    }
+
+    private boolean isChallengeQuestionsProvided(AuthenticatedUser user) {
+
+        try {
+            String userName = user.getUserName();
+            int tenantId = Utils.getTenantId(MultitenantUtils.getTenantDomain(userName));
+
+            UserStoreManager userStoreManager =
+                    FrameworkServiceDataHolder.getInstance().getRealmService()
+                            .getTenantUserRealm(tenantId)
+                            .getUserStoreManager();
+
+            Map<String, String> claimsMap = userStoreManager
+                    .getUserClaimValues(userName, new String[]{IdentityRecoveryConstants.CHALLENGE_QUESTION_URI},
+                            UserCoreConstants.DEFAULT_PROFILE);
+
+            String claimValue = claimsMap.get(IdentityRecoveryConstants.CHALLENGE_QUESTION_URI);
+            return claimValue != null;
+
+        } catch (IdentityException e) {
+            log.error("Identity exception occurred for user :" + user.getUserName(), e);
+
+        } catch (UserStoreException e) {
+            log.error("User store exception occurred for user :" + user.getUserName(), e);
+        }
+        return false;
+    }
+
+    private List<ChallengeQuestion> getChallengeQuestions(AuthenticatedUser user) {
+        String tenantDomain = MultitenantUtils.getTenantDomain(user.getUserName());
+
+        try {
+            return ChallengeQuestionManager.getInstance().getAllChallengeQuestions
+                    (tenantDomain);
+        } catch (IdentityRecoveryServerException e) {
+            log.error("Identity recovery server error occurred for user:" + user.getUserName(), e);
+            return null;
+        }
+    }
+
+    private void setChallengeQuestionAnswers(User user, UserChallengeAnswer[] userChallengeAnswers) {
+
+        try {
+            ChallengeQuestionManager.getInstance().setChallengesOfUser(user, userChallengeAnswers);
+        } catch (IdentityRecoveryException e) {
+            log.error("Unable to save challenge question answers", e);
+        }
+    }
+
+    private UserChallengeAnswer[] retrieveChallengeQuestionAnswers(HttpServletRequest servletRequest,
+                                                                   List<ChallengeQuestion> list) {
+        Map<String, String> questionsMap = new HashMap<>();
+        Map<String, String> answersMap = new HashMap<>();
+        List<UserChallengeAnswer> questionsAndAnswers = new ArrayList<>();
+        List<String> paramNamesList = new ArrayList<>();
+
+        Enumeration<String> paramNames = servletRequest.getParameterNames();
+        while (paramNames.hasMoreElements()) {
+            paramNamesList.add(paramNames.nextElement());
+        }
+
+        for (String requestParam : paramNamesList) {
+            if (requestParam.contains("Q-")) {
```

Comment:
Can we use constants for "Q-"

Reply:
yes. will use the existing constant for that